src/charm.py

```diff
@@ -1,104 +1,380 @@
 
-    https://discourse.charmhub.io/t/4208
-"""
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
 
+"""Charmed Machine Operator for OpenSearch."""
 import logging
+from datetime import datetime
+from typing import Dict, List, Optional
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+from charms.opensearch.v0.constants_tls import CertType
+from charms.opensearch.v0.helper_cluster import ClusterTopology, Node
+from charms.opensearch.v0.helper_databag import Scope
+from charms.opensearch.v0.helper_networking import units_ips
+from charms.opensearch.v0.helper_security import (
+    cert_expiration_remaining_hours,
+    generate_hashed_password,
+    to_pkcs8,
+)
+from charms.opensearch.v0.opensearch_base_charm import PEER, OpenSearchBaseCharm
+from charms.opensearch.v0.opensearch_distro import (
+    OpenSearchHttpError,
+    OpenSearchInstallError,
+    OpenSearchStartError,
+)
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateAvailableEvent,
+)
+from ops.charm import (
+    InstallEvent,
+    LeaderElectedEvent,
+    RelationChangedEvent,
+    RelationJoinedEvent,
+    StartEvent,
+    UpdateStatusEvent,
+)
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
 
-logger = logging.getLogger(__name__)
+from opensearch import OpenSearchTarball
 
+logger = logging.getLogger(__name__)
 
-class OperatorTemplateCharm(CharmBase):
-    """Charm the service."""
 
-    _stored = StoredState()
+class OpenSearchOperatorCharm(OpenSearchBaseCharm):
+    """This class represents the machine charm for OpenSearch."""
 
     def __init__(self, *args):
-        super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
+        super().__init__(*args, distro=OpenSearchTarball)  # OpenSearchSnap
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on.start, self._on_start)
+
+        self.framework.observe(self.on[PEER].relation_joined, self._on_peer_relation_joined)
+        self.framework.observe(self.on[PEER].relation_changed, self._on_peer_relation_changed)
+
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+    def _on_install(self, event: InstallEvent) -> None:
+        """Handle the install event."""
+        self.unit.status = MaintenanceStatus("Installing OpenSearch...")
+        try:
+            self.opensearch.install()
+        except OpenSearchInstallError:
+            self.unit.status = BlockedStatus("Could not install OpenSearch.")
+            event.defer()
```

Comment:
Are you deferring because a re-attempt can succeed without human operator intervention?

Reply:
Hmm, I thought I would handle the case where the HTTP request to download the binaries fails. Maybe I should attempt a couple of retries on the HTTP download and if it fails again, put it into BlockedStatus? What do you think?

Reply:
I think that's a good approach. Most of the data platform charms are using the `tenacity` library to accomplish that, so it can be useful here too.